python/src/equistore/operations/reduce.py

```diff
@@ -0,0 +1,99 @@
+import numpy as np
+
+
+from equistore import Labels, TensorBlock, TensorMap
+
+from . import _dispatch
+
+# TODO maybe add a way to reduce only the rows
+#  with sample[sample_name] = idx
+# for example not summing over all the sample
+# that have the same 'structure'
+# but only those with 'structure' = 0
+
+
+def _reduction_sample_block(block: TensorBlock, sample_name: list) -> TensorBlock:
+    """Create a new :py:class:`TensorBlock` summing the `properties` among
+    the selected `samples`.
+    The output :py:class:`TensorBlocks` have the same components of the input one.
+    :param block: :py:class:`TensorBlock` -> input block
+    :param sample_name: list -> names of samples to sum
+    """
+    for isample in sample_name:
+        if isample not in block.samples.names:
+            raise ValueError(
+                "the values in sample_names \
+                             should be contained in block1.samples.names"
+            )
+
+    # get the index of the selected sample
+    sample_selected = [block.samples.names.index(isample) for isample in sample_name]
+    # create a dictionary with the lists of all the index with the same sample_selected
+    samples = block.samples.view(dtype=np.int32).reshape(block.samples.shape[0], -1)
+    s, index = np.unique(samples[:, sample_selected], return_inverse=True, axis=0)
+    values_result = _dispatch.index_add(block.values, s.shape[0], index)
+
+    # if s.size == 1:
+    #    s = s.reshape(1, len(s))
+
+    result_block = TensorBlock(
+        values=values_result,
+        samples=Labels(
+            [i for i in sample_name], np.array([i for i in s], dtype=np.int32)
+        ),  # TODO guardare
+        components=block.components,
+        properties=block.properties,
+    )
+
+    for parameter, gradient in block.gradients():
+        # here I copy because I want to modify the samples array
+        samples = (
+            gradient.samples.view(dtype=np.int32)
+            .reshape(gradient.samples.shape[0], -1)
+            .copy()
+        )
+        for i in range(len(samples)):
+            samples[i, 0] = index[samples[i, 0]]
```

Comment:
is there a magical numpy function that would do this without a for loop?

Reply:
it think samples[:, 0] = index[samples[:, 0]] works